### DISCUSSION TITLE
Interceptors遇到的两个疑难问题，还请百忙之中帮忙看看是不是我哪里用得不对，谢谢

### DISCUSSION THREAD

**Discussion by Selol** [dotnetcore/EasyCaching#448] [Q&A]
Hi, 
我在测试Interceptors的时候发现两个问题，还请百忙之中帮忙看看是否是我那里用的不对，多谢。

问题1：easycachingable的expiration=5时实际要更久才能刷新
复现步骤：
1. 用sample的代码 https://github.com/dotnetcore/EasyCaching/tree/dev/sample/EasyCaching.Demo.Interceptors
2. 尝试用castle方式，修改代码如下图
![image](https://user-images.githubusercontent.com/11731277/215483543-6676dead-fab0-46ef-bd17-874ff7f68f61.png)

3. 启动项目
预期行为：https://localhost:54738/api/values/castle的返回值应该5秒一变
实际：可能要几分钟才变，见下图

![2](https://user-images.githubusercontent.com/11731277/215474401-d5e9a855-e2c6-435d-a794-7f26115cf10f.gif)

问题2：builder.ConfigureCastleInterceptor()放在最后会忽略之前的builder.RegisterType，放在前面则cache不生效。同时builder.ConfigureCastleInterceptor()会忽略service.AddScoped等方法

环境：.net6, .netcore 3.1
复现步骤
1. WeatherSerivce和AnotherService均实现TestInterface
2. 代码如下
Program.cs
```c#
using Autofac;
using Autofac.Extensions.DependencyInjection;
using Autofac.Extras.DynamicProxy;
using EasyCaching.Interceptor.Castle;
using MyWebApi.Services;

var builder = WebApplication.CreateBuilder(args);

builder.Host.UseServiceProviderFactory(new AutofacServiceProviderFactory())
    .ConfigureContainer<ContainerBuilder>(
        builder =>
        {
            //builder.RegisterType<WeatherSerivce>().As<TestInterface>().EnableInterfaceInterceptors();
            builder.RegisterType<AnotherService>().As<TestInterface>();
            builder.ConfigureCastleInterceptor();
        })
    .ConfigureServices((services) =>
    {
        //services.AddScoped<TestInterface, WeatherSerivce>();
        services.AddMemoryCache();
        services.AddEasyCaching(options => options.UseInMemory("m1"));
        services.ConfigureCastleInterceptor(options => options.CacheProviderName = "m1");
    });



// Add services to the container.
//builder.Services.AddEasyCaching(options => { options.UseInMemory("m1"); });
//builder.Services.ConfigureCastleInterceptor(options => options.CacheProviderName = "m1");
//builder.Services.AddScoped<TestInterface, WeatherSerivce>();

builder.Services.AddControllers();
// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

//builder.Services.AddSingleton<WeatherSerivce, WeatherSerivce>();
builder.Services.AddMvc().AddControllersAsServices();

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();

app.UseAuthorization();

app.MapControllers();

app.Run();

```
Controller:
![image](https://user-images.githubusercontent.com/11731277/215480633-13dd074a-b1ae-4c44-813d-20af23ac03bf.png)

尝试访问该接口，发现被注入的service不是指定的AnotherService而是WeatherService
![image](https://user-images.githubusercontent.com/11731277/215481741-8448c84b-18cc-465f-a6a9-d7f2b808ef82.png)
如果交换ConfigureCastleInterceptor和RegisterType的顺序，被注入的service正确但并未被castle正确拦截
![image](https://user-images.githubusercontent.com/11731277/215481338-40d39204-d317-4ced-98c3-c2233e184061.png)

.netcore 3.1无法人为指定顺序，但指定的依赖绑定失效。

update:
似乎找到了一种符合预期的写法，但不太确定是否是best practice
```c#
builder.RegisterType<EasyCachingInterceptor>();
builder.RegisterType<AnotherService>().As<TestInterface>().EnableInterfaceInterceptors().InterceptedBy(typeof(EasyCachingInterceptor));
```

**Comment by catcherwong** [comment]
> 问题1：easycachingable的expiration=5时实际要更久才能刷新

默认情况，设置缓存的时候会加一个随机过期时间，打散过期时间，避免集中过期，如果不需要这个，可以设置 maxrdsecond 为 0。

```cs
option.UseInMemory(x => 
{
    x.MaxRdSecond = 0;
}, "default");
```

**Reply by Selol** [reply]
非常感谢您的回复，请问问题2我应该怎么处理呢，多谢？
